### PR TITLE
Add option to skip cudagraphing on dynamic shape graphs

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122520

This was requested internally.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

Differential Revision: [D55264528](https://our.internmc.facebook.com/intern/diff/D55264528)

### CODE REVIEW

File: torch/_inductor/config.py [hunk 1]
```diff
@@ -531,6 +531,10 @@ class triton:
     # Use cudagraph trees for memory pooling if `cudagraphs` is True
     cudagraph_trees = True
 
+    # Should we skip cudagraphing graphs with dynamic shape inputs
+    # If False, we will re-record a graph for each unique set of shape inputs
+    skip_dynamic_graphs = True
```
Comment (mlazos) [inline torch/_inductor/config.py:536]: nit: should we say cudagraphs_skip_dynamic_graphs

Comment (eellison) [reply torch/_inductor/config.py:536]: yea, agreed

File: test/inductor/test_cudagraph_trees.py [hunk 2]
```diff
@@ -269,6 +269,19 @@ def foo(x, y):
                 captured_output[0]
             )
 
+        @torch._inductor.config.patch("triton.skip_dynamic_graphs", True)
```
Comment (mlazos) [inline test/inductor/test_cudagraph_trees.py:272]: Should you also test that multiple compiles with different shapes still return correct results? or is that the status quo

**Review by mlazos — APPROVED** [review]

File: test/inductor/test_cudagraph_trees.py (hunk 2 above)
Comment (eellison) [reply test/inductor/test_cudagraph_trees.py]: already tested: https://github.com/pytorch/pytorch/commit/10cbc9a063d20116ded39fb54a59db9d17e53acf#diff-70ac3047d6e27843ff4f81ee908669839ca9551814ccbd47b2c3354997f87c73R1123-R1180

**Comment by ezyang** [comment]
You changed the default, which is fine, but then you need to flip it back for the benchmark

File: test/inductor/test_cudagraph_trees.py [hunk 3]
```diff
@@ -269,6 +269,19 @@ def foo(x, y):
                 captured_output[0]
             )
 
+        @torch._inductor.config.patch("triton.cudagraph_skip_dynamic_graphs", True)
```
Comment (mlazos) [inline test/inductor/test_cudagraph_trees.py:272]: Should you add an addional test that we properly re-record the shape or is that already covered?

**Comment by eellison** [comment]
@ezyang oops good catch

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55264528).

File: torch/_inductor/compile_fx.py [hunk 4]
```diff
@@ -727,6 +727,31 @@ def fx_codegen_and_compile(
...
+                    if (
+                        node.op == "placeholder"
+                        or not isinstance(meta_val, torch.Tensor)
+                        or not torch._inductor.utils.any_is_symbolic(meta_val)
+                    ):
+                        continue
+
+                    if stack_trace := node.meta.get("stack_trace", None):
+                        break
+                disable = "graph with symbolic shapes inputs and config.triton.skip_dynamic_graphs=True."
```
Comment (shunting314) [inline torch/_inductor/compile_fx.py:748]: typo on the config name?

**Review by shunting314 — APPROVED** [review]

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55264528).

**Pull request closed without merging**